examples/restaurant/eval/test_agent.py

```diff
@@ -0,0 +1,168 @@
+        if sess.current_speech:
+            # `generate_reply` is called in `on_enter`
+            await sess.current_speech
+
+        result = await sess.run(user_input="I want to order a pizza and a salad")
+        try:
+            result.expect[:].contains_function_call(name="to_takeaway")
+        except AssertionError:
+            result = await sess.run(user_input="okay")
+        result.expect.maybe_message(role="assistant")
+        result.expect.function_call(name="to_takeaway")
+        result.expect.function_call_output()
+        result.expect.maybe_message(role="assistant")
+        result.expect.agent_handoff(new_agent_type=Takeaway)
+
+        result = await sess.run(user_input="I want to add a coffee")
+        try:
+            result.expect[:].contains_function_call(name="update_order")
+        except AssertionError:
+            result = await sess.run(user_input="yes, that's all I want")
+        result.expect.maybe_message(role="assistant")
+        result.expect.function_call(name="update_order")
+        result.expect.function_call_output()
+        try:
+            result.expect[:].contains_function_call(name="to_checkout")
+        except AssertionError:
+            result = await sess.run(user_input="yes, I want to checkout")
+
+        result.expect.maybe_message(role="assistant")
+        result.expect.function_call(name="to_checkout")
+        result.expect.function_call_output()
+        result.expect.maybe_message(role="assistant")
+        result.expect.agent_handoff(new_agent_type=Checkout)
+        await result.expect.message(role="assistant").judge(
+            llm, intent="must tell user the expense is $17", verbose=True
+        )
+
+        # test checkout
+        checkout_chat_ctx = agents["checkout"].chat_ctx.copy()
+
+        # branch 1: user disagree with the expense
+        await sess.current_agent.update_chat_ctx(checkout_chat_ctx)
+        result = await sess.run(user_input="no, it should be $16")
+        try:
+            result.expect[:].contains_function_call(name="to_takeaway")  # back to takeaway
+        except AssertionError:
+            await result.expect.message(role="assistant").judge(
+                llm,
+                intent="must disagree with the $16 or ask user to update the order",
+                verbose=True,
+            )
+
+        # branch 2: user wants to change the order
+        if sess.current_agent != agents["checkout"]:
+            await sess._update_activity(agents["checkout"], new_activity="start")
+        await sess.current_agent.update_chat_ctx(checkout_chat_ctx)
+        result = await sess.run(user_input="oh I want to add another coffee")
+        try:
+            result.expect[:].contains_function_call(name="to_takeaway")
+        except AssertionError:
+            result.expect.message(role="assistant").judge(
+                llm, intent="ask user if they want to back to takeaway", verbose=True
+            )
```

Comment:
I think we need to think about a better way to do this. I think it will not be possible for v1 (at least cleanly)